File: templates/ts-node-aoc/lib/get-aoc-input/scaffold.ts
import fs from 'node:fs';
import path from 'node:path';

interface WriteTaskInputPayload {
  data: string;
  outputDir: string;
}
export function writeAOCTaskinput({ data, outputDir }: WriteTaskInputPayload): void {
  const taskInputPath = path.join(outputDir, 'input.txt');
  return fs.writeFileSync(taskInputPath, data);
}

interface ScaffoldTemplatePayload {
  year: number;
  day: number;
  outputDir: string;
  title: string;
}
export function scaffoldAOCTemplate({ year, day, outputDir, title }: ScaffoldTemplatePayload): void {
  const README_TEMPLATE = `[${title}](https://adventofcode.com/${year}/day/${day} "${title}")

\`\`\`shell
npx oxnode task.ts
\`\`\`
`;
  const readmePath = path.join(outputDir, 'README.md');
  fs.writeFileSync(readmePath, README_TEMPLATE);

  const TASK_TEMPLATE = `import { parseInput } from '../../lib/parse-input';

function part1(input: string[]) {}

function part2(input: string[]) {}

function go(): void {
  console.time('task');

  console.time('parse-input');
  const input = parseInput('./input.txt');
  console.timeEnd('parse-input');

  console.time('part 1');
  const res1 = part1(input);
  console.log('PART 1: ', res1);
  console.timeEnd('part 1');

  console.time('part 2');
  const res2 = part2(input);
  console.log('PART 2: ', res2);
  console.timeEnd('part 2');

  console.timeEnd('task');
}

go();
`;
  const taskPath = path.join(outputDir, 'task.ts');
  fs.writeFileSync(taskPath, TASK_TEMPLATE);
}

File: templates/rust-aoc/lib/get_aoc_input/scaffold.rs
use std::fs;

pub struct WriteTaskInput<'a> {
    pub output_dir: &'a str,
    pub data: &'a str,
}

pub fn write_task(input: WriteTaskInput) -> std::io::Result<()> {
    fs::write(format!("{}/input.txt", input.output_dir), input.data)
}

pub struct WriteTemplateInput<'a> {
    pub output_dir: &'a str,
    pub year: u32,
    pub day: u32,
    pub title: &'a str,
}

pub fn write_template(input: WriteTemplateInput) -> std::io::Result<()> {
    let WriteTemplateInput {
        output_dir,
        year,
        day,
        title,
    } = input;
    let url = format!("https://adventofcode.com/{year}/day/{day}");
    let readme =
        format!("[{title}]({url} \"{title}\")\n\n```shell\ncargo run --bin day{day}\n```\n");
    fs::write(format!("{output_dir}/README.md"), readme)?;

    let skeleton = "use std::time::Instant;\n\n\
         fn part1(_lines: &[String]) -> i32 {\n\
         \t0\n\
         }\n\n\
         fn part2(_lines: &[String]) -> i32 {\n\
         \t0\n\
         }\n\n\
         fn main() {\n\
         \tlet start = Instant::now();\n\
         \tlet lines = aoc_cli::parse_input::lines(\"./input.txt\");\n\
         \tlet p1 = part1(&lines);\n\
         \tprintln!(\"PART 1: {p1}\");\n\
         \tlet p2 = part2(&lines);\n\
         \tprintln!(\"PART 2: {p2}\");\n\
         \tprintln!(\"Elapsed: {:?}\", start.elapsed());\n\
         }\n";
    fs::create_dir_all(format!("{output_dir}/src/bin"))?;
    fs::write(format!("{output_dir}/src/bin/day{day}.rs"), skeleton)
}

pub fn register_bin(day: u32) -> std::io::Result<()> {
    let entry =
        format!("\n[[bin]]\nname = \"day{day}\"\npath = \"tasks/day{day}/src/bin/day{day}.rs\"\n");
    fs::OpenOptions::new()
        .append(true)
        .open("Cargo.toml")
        .and_then(|mut f| std::io::Write::write_all(&mut f, entry.as_bytes()))
}
